src/fs/metadata_log_operations/create_file.hh

```diff
@@ -0,0 +1,127 @@
+/*
+ * This file is open source software, licensed to you under the terms
+ * of the Apache License, Version 2.0 (the "License").  See the NOTICE file
+ * distributed with this work for additional information regarding copyright
+ * ownership.  You may not use this file except in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+/*
+ * Copyright (C) 2020 ScyllaDB
+ */
+
+#pragma once
+
+#include "fs/metadata_log.hh"
+#include "fs/path.hh"
+
+namespace seastar::fs {
+
+class create_file_operation {
+    metadata_log& _metadata_log;
+    bool _is_directory;
+    sstring _entry_name;
+    file_permissions _perms;
+    inode_t _dir_inode;
+    inode_info::directory* _dir_info;
+    ondisk_create_inode_as_dir_entry_header _ondisk_entry;
+
+    create_file_operation(metadata_log& metadata_log) : _metadata_log(metadata_log) {}
+
+    future<inode_t> perform_1(sstring path, file_permissions perms, bool is_directory) {
+        _is_directory = is_directory;
+        if (is_directory) {
+            while (not path.empty() and path.back() == '/') {
+                path.erase(path.end() - 1, path.end());
+            }
+        }
+
+        _entry_name = last_component(path);
+        if (_entry_name.empty()) {
+            if (is_directory) {
+                return make_exception_future<inode_t>(std::runtime_error("Invalid path"));
+            } else {
+                return make_exception_future<inode_t>(std::runtime_error("Path has to end with character different than '/'"));
+            }
+        }
+
+        _perms = perms;
+        path.erase(path.end() - _entry_name.size(), path.end());
+        return _metadata_log.futurized_path_lookup(path).then([this](inode_t dir_inode) {
+            _dir_inode = dir_inode;
+            return with_semaphore(_metadata_log._create_or_delete_lock, 1, [this] {
+                // We do not have to shared lock dir_inode because we hold global lock
+                return perform_2();
+            });
+        });
+    }
+
+    future<inode_t> perform_2() {
+        auto dir_it = _metadata_log._inodes.find(_dir_inode);
+        if (dir_it == _metadata_log._inodes.end()) {
+            return make_exception_future<inode_t>(operation_became_invalid_exception());
+        }
+
+        assert(dir_it->second.is_directory());
+        _dir_info = &dir_it->second.get_directory();
+
+        return _metadata_log._dir_entry_locks.with_lock_on({_dir_inode, _entry_name}, [this] {
+            return perform_3();
+        });
+    }
+
+    future<inode_t> perform_3() {
```

Comment:
That's the third worst function name I've ever seen, and that's only because I scrolled over `perform_1` and `perform_2` first :smile: But, back to being serious, please add descriptive names - they may be long, but they should express what the function does.

Reply:
The problem with naming here is that the flow is somewhat inverted. perform_1 returns future resolving when whole operation is complete. perform_2 returns the same, but assumes that path parsing is done and global lock is acquired. perform_3 returns the same, but assumes what perform_2 assumed and also that an unique lock is acquired on the directory entry to create. I don't have a good idea how to name it, maybe as if the flow was not inverted?

Reply:
It's fine for the order to be inverted and mixed, it's quite common for seastar apps. Let's just name functions based on what they do - if `perform_1` is the main entry point, so `create_file` sounds a little obvious, but it's still a good name. `perform_2` assumes we looked up a dir inode, so something along `create_file_in_directory` will explain its purpose. Finally, `create_file_in_directory` is just a thin wrapper to get the directory and its lock, so `perform_3` can be named `do_create_file_in_directory`, since that's what actually happens in there.

`perform_2` and `perform_3` are only used internally, so the users wouldn't care if their order is inverted or not - but it's important to name them in a self-explanatory way, so that reviewers and developers know what was the intent.